Adding customize block placement (#6789)
magento/devdocs: src/recommendations/customize.md

---
group: product-recommendations
title: Customize Recommendations
---

When you install the Product Recommendations module, Magento creates the `ProductRecommendationsLayout` directory under your Magento theme directory. This directory contains files that you can customize to change how the recommendations appear on your storefront and where those recommendation units are placed.

The following table describes each subdirectory under `ProductRecommendationsLayout`:

|**Directory**|**Purpose**|
|---|---|
|`layout`|Contains `*.xml` files for each page type|
|`templates`|Contains files that call the fetch and render scripts|

> |`web\js`|Contains the JavaScript files that fetch and render recommendations for your store|

Reviewer: Forward slashes for directory names `web/js`
```suggestion
|`web/js`|Contains the JavaScript files that fetch and render recommendations for your store|
```

> |`web\template`|Contains the template for the `magento/product-recommendations` module|

Reviewer: ```suggestion
|`web/template`|Contains the template for the `magento/product-recommendations` module|
```

## Recommendation unit placement

When you [create a new recommendation](https://docs.magento.com/m2/ee/user_guide/marketing/create-new-rec.html), you specify [where you want the recommendation unit to appear](https://docs.magento.com/m2/ee/user_guide/marketing/product-recommendations.html#productrecplacement) on your page. You can select either above the main content or below the main content. You can, however, customize that location by editing the `*.xml` files responsible for the layout. Magento creates these files when you create a recommendation.

1. Locate the `layout` directory: `<your theme>/Magento_ProductRecommendationsLayout/layout/<page handles>.xml`.

    The following table lists the XML filenames used in the above `<page handles>` variable depending on your page:

    |Filename|Page|
    |---|---|
    |`catalog_category_view.xml`|Category|
    |`catalog_product_view.xml`|Product detail|
    |`checkout_cart_index.xml`|Cart|
    |`checkout_onepage_success.xml`|Checkout|
    |`cms_index_index.xml`|Home|

> 1. Open the page you want to modify. For example, let's modify the `catalog_category_view.xml` file. It should look something like this:

Reviewer 2: I would change this to catalog_product_view.xml..in other words the product detail page because we are giving them the example of changes to the product detail page

        <referenceBlock name="page.wrapper">
            <block class="Magento\Framework\View\Element\Template" before="-" name="product_recommendations_fetcher" template="Magento_ProductRecommendationsStorefront::fetcher.phtml" />
        </referenceBlock>
        <body>
            <referenceBlock name="main.content">
                <block class="Magento\ProductRecommendationsStorefront\Block\Renderer" after="-" name="product_recommendations_product_below_content" template="Magento_ProductRecommendationsStorefront::renderer.phtml">
                    <arguments>
                        <argument name="pagePlacement" xsi:type="string">below-main-content</argument>
                    </arguments>
                </block>
            </referenceBlock>
        </body>
    </page>
    ```

    Note that `<referenceBlock name="main.content">` contains `name="product_recommendations_product_below_content"` which means the recommendation unit will appear after the main content. But maybe you want the recommendation unit to appear after the product image on the product detail page.

> 1. Here is the code after you make your updates:

Reviewer: 1. Replace <original block code> with <new block code>.

Rewriting this step in this manner will avoid the "note this has changed" text afterward.

|`<a data-bind="attr : {'data-sku' : sku, 'data-unit-id'}"...</a>`| Collects click events. <br/>**Note:** If you add any anchor tags, you must include these attributes.|

Refer to the [template overview]({{ site.baseurl }}/guides/v{{ site.version }}/frontend-dev-guide/templates/template-walkthrough.html) for more information about how to modify templates in Magento.

## Recommendation unit placement

When you [create a recommendation](https://docs.magento.com/m2/ee/user_guide/marketing/create-new-rec.html), you specify [where you want the recommendation unit to appear](https://docs.magento.com/m2/ee/user_guide/marketing/product-recommendations.html#productrecplacement) on your page. In the Admin, you can select to place the recommendations either above or below the main content. You can, however, customize that location by editing the `*.xml` files responsible for the layout. Magento generates these files when you create a recommendation.

1. Change to the `layout` directory:

    ```bash
    cd `<your theme>/ProductRecommendationsLayout/layout`
    ```

>     Within this directory you will see several `<page handles>.xml` files

Reviewer: Delete this line. The previous revision made that page handles variable unnecessary.

    ```xml
    <?xml version="1.0"?>
    <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
        <referenceBlock name="page.wrapper">
            <block class="Magento\Framework\View\Element\Template" before="-" name="product_recommendations_fetcher" template="Magento_ProductRecommendationsStorefront::fetcher.phtml" />
        </referenceBlock>
        <body>
            <referenceBlock name="main.content">
                <block class="Magento\ProductRecommendationsStorefront\Block\Renderer" after="-" name="product_recommendations_product_below_content" template="Magento_ProductRecommendationsStorefront::renderer.phtml">
                    <arguments>
                        <argument name="pagePlacement" xsi:type="string">below-main-content</argument>
                    </arguments>
                </block>
            </referenceBlock>
        </body>
    </page>
    ```

>     In the above snippet, the reference block `name` specifies `main.content`, which defines the block on the page. The block `class` contains the `after="-"` attribute, which indicates that the recommendation unit will be displayed on the page after the main content block.

Reviewer: ```suggestion
    In the above snippet, the `main.content` reference block indicates the recommendation unit will be placed somewhere relative to that element. Its `block` element contains the `after="-"` attribute, which indicates that the recommendation unit will be displayed on the page after the main content block.
```

1. Let's modify this file by specifying a different content block:

    ```xml
    <?xml version="1.0"?>
    <page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
        <referenceBlock name="page.wrapper">
            <block class="Magento\Framework\View\Element\Template" before="-" name="product_recommendations_fetcher" template="Magento_ProductRecommendationsStorefront::fetcher.phtml" />
        </referenceBlock>
        <body>
            <referenceBlock name="product.info.media">
                <block class="Magento\ProductRecommendationsStorefront\Block\Renderer" after="-" name="product_recommendations_product_below_content" template="Magento_ProductRecommendationsStorefront::renderer.phtml">
                    <arguments>
                        <argument name="pagePlacement" xsi:type="string">below-main-content</argument>
                    </arguments>
                </block>
            </referenceBlock>
        </body>
    </page>
    ```

    In the above modified snippet, the reference block `name` changes from `main.content` to `product.info.media`.

>     This change results in your recommendation unit appearing after the product image on the product detail page. You can specify any open-source container layout name. You can also change the location of the recommendation unit to appear before or after the block by specifying the `after="-"` or `before="-"` attribute.

Reviewer: Move the first sentence into the previous paragraph. Start a new paragraph with "You".

I don't think it's necessary to mention open source. Possibly the merchant has installed/created custom templates.

Also, why is the value for the `pagePlacement` argument below-main-content the same for both examples? And if you places the recs above an element, will the value be something else?
